cf-testd/cf-testd.c

```diff
@@ -0,0 +1,492 @@
+    /* TODO: why do we do this ?  We fail if getsockname() fails, but
+     * don't use the information it returned.  Was the intent to use
+     * it to fill in conn->ipaddr ? */
+    struct sockaddr_storage addr;
+    socklen_t size = sizeof(addr);
+
+    if (getsockname(ConnectionInfoSocket(info), (struct sockaddr *)&addr, &size) == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Could not obtain socket address. (getsockname: '%s')",
+            GetErrorStr());
+        return NULL;
+    }
+#endif
+
+    ServerConnectionState *conn = xcalloc(1, sizeof(*conn));
+    conn->ctx = NULL;
+    conn->conn_info = info;
+    conn->encryption_type = 'c';
+    /* Only public files (chmod o+r) accessible to non-root */
+    conn->uid = CF_UNKNOWN_OWNER;                    /* Careful, 0 is root! */
+    /* conn->maproot is false: only public files (chmod o+r) are accessible */
+
+    Log(LOG_LEVEL_DEBUG,
+        "New connection (socket %d).",
+        ConnectionInfoSocket(info));
+    return conn;
+}
+
+static void CFTestD_DeleteConn(ServerConnectionState *conn)
+{
+    int sd = ConnectionInfoSocket(conn->conn_info);
+    if (sd != SOCKET_INVALID)
+    {
+        cf_closesocket(sd);
+    }
+    ConnectionInfoDestroy(&conn->conn_info);
+
+    *conn = (ServerConnectionState) {0};
+    free(conn->session_key);
+    free(conn);
+}
+
+static void *CFTestD_HandleConnection(void *c)
+{
+    ServerConnectionState *conn = c;
+
+    char aligned_ipaddr[CF_MAX_IP_LEN + 2];
+
+    strlcpy(aligned_ipaddr, conn->ipaddr, sizeof(aligned_ipaddr));
+    strlcat(aligned_ipaddr, "> ",         sizeof(aligned_ipaddr));
+    /* Pad with enough spaces for IPv4 addresses to be aligned. Max chars are
+     * 15 for the address plus two for "> " == 17. */
+    size_t len;
+    for (len = strlen(aligned_ipaddr); len < 17; len++)
+    {
+        aligned_ipaddr[len] = ' ';
+    }
+    aligned_ipaddr[len] = '\0';
+
+
+    Log(LOG_LEVEL_INFO, "Accepting connection");
+
+    bool established = CFTestD_TLSSessionEstablish(conn);
+    if (!established)
+    {
+        Log(LOG_LEVEL_ERR, "Could not establish TLS Session");
+        return NULL;
+    }
+    int ret = getnameinfo((const struct sockaddr *) &conn->conn_info->ss,
+                      conn->conn_info->ss_len,
+                      conn->revdns, sizeof(conn->revdns),
+                      NULL, 0, NI_NAMEREQD);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_INFO,
+            "Reverse lookup failed (getnameinfo: %s)!",
+            gai_strerror(ret));
+    }
+    else
+    {
+        Log(LOG_LEVEL_INFO,
+            "Hostname (reverse looked up): %s",
+            conn->revdns);
+    }
+
+    while (CFTestD_BusyLoop(conn))
+    {
+    }
+    /* ============================================================ */
+
+    CFTestD_DeleteConn(conn);
+    return NULL;
+}
+
+static void CFTestD_SpawnConnection(const char *ipaddr, ConnectionInfo *info)
+{
+    ServerConnectionState *conn = NULL;
+    conn = CFTestD_NewConn(info);
```

Comment:
I think these two lines can be squashed. Or is the combination way too long?

Reply:
Thanks, fixed.